### ISSUE TITLE
Add `keyValueIterator` field to everything

### ISSUE THREAD

**Issue by Simn** [HaxeFoundation/haxe#7395]
* [ ] Array (@simn)
* [x] haxe.ds.BalancedTree  (@markknol)
* [x] haxe.ds.EnumValueMap  (@markknol)
* [x] haxe.ds.IntMap  (@markknol)
* [x] haxe.ds.List (lol) (@markknol)
* [x] haxe.ds.Map (@markknol)
* [x] haxe.ds.ObjectMap (@markknol)
* [x] haxe.ds.StringMap (@markknol)
* [ ] haxe.ds.Vector
* [x] haxe.ds.WeakMap (@markknol)
* [x] String (@realyuniquename)
* [ ] Whatever I'm forgetting

Community effort?

**Comment by markknol** [comment]
Is there a  example of one implementation?

**Comment by Simn** [comment]
At the moment only in the tests: https://github.com/HaxeFoundation/haxe/blob/development/tests/unit/src/unit/TestKeyValueIterator.hx#L17

It works just like `iterator`, except that it returns `{key: K, value: V}`.

**Comment by nadako** [comment]
So I was looking into implementing `KeyValueIterator` for `Array` and a couple questions arose:

 - Do we want a common type for it across all targets (e.g. `ArrayKeyValueIterator<T>`), that we would return instead of a general `KeyValueIterator<Int,T>` for better inlining? If so, should it be in the `Array` module? 
   - If we do this, maybe we should do the same for `ArrayIterator` so we fix #5981?
 - I guess we want to make `for (k=>v in (array:KeyValueIterable<Int,...>)` working on e.g. JS via the same hacks as we do for `Iterable`?
   - By the way, `(for k=>v in (a:Dynamic))` is not a compile-time error (it treats `a` as `Iterator`), while normal `for (v in (a:Dynamic))` gives `You can't iterate on a Dynamic value, please specify Iterator or Iterable` error.

**Comment by Simn** [comment]
> Do we want a common type for it across all targets (e.g. ArrayKeyValueIterator<T>), that we would return instead of a general KeyValueIterator<Int,T> for better inlining? If so, should it be in the Array module?

That's a good idea because I don't have a solution for #7379.

> If we do this, maybe we should do the same for ArrayIterator so we fix #5981?

Makes sense

> I guess we want to make for (k=>v in (array:KeyValueIterable<Int,...>) working on e.g. JS via the same hacks as we do for Iterable?

Yes I think that's necessary.

> By the way, (for k=>v in (a:Dynamic)) is not a compile-time error (it treats a as Iterator), while normal for (v in (a:Dynamic)) gives You can't iterate on a Dynamic value, please specify Iterator or Iterable error.

Ah yeah, will fix!

**Comment by Simn** [comment]
> Ah yeah, will fix!

Done

**Comment by RealyUniqueName** [comment]
Is it possible to make key-value iterator on `haxe.ds.Map` inlinable? Or there is the same problem as with `Map.keys()` and `Map.iterator()`?

**Comment by markknol** [comment]
Can we add this ArrayKeyValueIterator for now?

```haxe
class ArrayKeyValueIterator<T>{
	var index:Int = 0;
	var array:Array<T>;

	public inline function new(array:Array<T>) {
		this.array = array;
	}
	
	public inline function hasNext() return index < array.length;
	
	public inline function next() {
		var val = { key: index, value: array[index] };
		index ++;
		return val;
	}
}
```

**Comment by RealyUniqueName** [comment]
String is implemented in https://github.com/HaxeFoundation/haxe/pull/8331

**Comment by kLabz** [comment]
> String is implemented in #8331

Note that one needs `using StringTools` to be able to use it though

**Comment by markknol** [comment]
### Proposal
Can we have this in the std?

```haxe
package haxe.iterators;

/**
	Support for `for(index => value in array)` syntax.
**/
final class ArrayKeyValueIterator<T> {
	var index:Int = 0;
	final array:Array<T>;

	public inline function new(array:Array<T>) {
		this.array = array;
	}

	public inline function hasNext() return index < array.length;

	public inline function next() {
		return  { key: index, value: array[index ++] };
	}

	public static inline function keyValueIterator<A>(arr:Array<A>):ArrayKeyValueIterator<A> {
		return new ArrayKeyValueIterator(arr);
	}
}
```
### Usage

Add to imports / [import.hx](https://haxe.org/manual/type-system-import-defaults.html)

```haxe
using haxe.iterators.ArrayKeyValueIterator;
```

Then we can to do just this:

```haxe
for (idx => value in ["apple", "banana", "pear"]) {
	trace(idx, value);
}
```

I have this in almost all of my projects, but I feel like its so simple that it should be part of the std.

**Comment by Gama11** [comment]
@markknol You won't need any `using` at all once https://github.com/HaxeFoundation/haxe/pull/7422 is done.

**Comment by Simn** [comment]
I think we can close this. If anyone misses an implementation on a specific type they can open a new issue for it.
